### PR TITLE
Add support for advanced ${} uses

### LINKED ISSUE #8667: It's difficult to correctly construct values for common environment variables
Environment variables like `PATH`, `LD_LIBRARY_PATH`, `PYTHONPATH ` combine values that are usually contributed from many different pieces of software.

For example, one image might install a Python library to /opt/acme and set `PYTHONPATH` to this value:

    ENV PYTHONPATH /opt/acme

An image based on this one might install another Python library in /usr/local/lib/yoyodyne.  It might try to cooperate with the underlying image:

    ENV PYTHONPATH $PYTHONPATH:/usr/local/lib/yoyodyne

This may prove correct, yielding a value for the variable in the container of `/opt/acme:/usr/local/lib/yoyodyne`.  However, it requires the author of the second image to know about the construction of the first image.  It also requires them to keep themselves updated about changes to the construction of the first image.  If a future version changes the way the software is installed so that `PYTHONPATH` no longer need be set then unless the Dockerfile for the second image is updated, the value for `PYTHONPATH` becomes `:/usr/local/lib/yoyodyne` (notice the leading `:`).  This mostly works but it introduces an extra unintended path to the Python module search path.  This may not cause any problems - or it might break the image (in a worst case outcome, perhaps it breaks the image quietly, by leaving everything with the appearance of working but introducing a subtle security issue where the Python application packaged in the second image begins evaluating arbitrary user-supplied Python code).

If full shell-style variable substitution were supported then this could be avoided by a diligent Dockerfile author:

    ENV PYTHONPATH ${PYTHONPATH:+${PYTHONPATH}:}/usr/local/lib/yoyodyne

The first part of this says "If `PYTHONPATH` is set, substitute its value followed by `:`.  If it is not set, substitute the empty string".  The result of this substitution is then concatenated with the literal `/usr/local/lib/yoyodyne`.  This produces either `/opt/acme:/usr/local/lib/yoyodyne` if `PYTHONPATH` is already set to `/opt/acme` or `/usr/local/lib/yoyodyne` if `PYTHONPATH` is not already set.

There are probably other solutions to this problem too.  The more complicated shell substitution is mostly given as an example of how some existing systems account for this problem and offer a solution for it.

### PR DESCRIPTION
**Pull request by duglin**
In particular ${xxx:+...} and ${xxx:-...}

~~This depends on https://github.com/docker/docker/pull/10431~~

Closes #8667

### CODE REVIEW

**Comment by moxiegirl** [comment]
LGTM

**Comment by duglin** [comment]
Interesting usecase: https://github.com/docker/docker/issues/9731

**Comment by ahmetalpbalkan** [comment]
94379ae and 76ac2e3 have the same commit message. can you squash/edit?

**Comment by tiborvass** [comment]
I'm closing this until we merge the dependency. We can reopen after :)

**Comment by duglin** [comment]
Huh?  Is that really the normal process?  Seems odd to me.

**Comment by tiborvass** [comment]
@duglin yes :) We actually had a discussion this morning about PR dependencies, and we do this every now and then, we may not be enforcing it well enough.

**Comment by duglin** [comment]
Obviously that means the other PR needs to be approved. ☺

-Doug

**Comment by icecrime** [comment]
Moving this in `1-design-review`.

I like the feature, and the implementation is simple enough, so +1.

**Comment by duglin** [comment]
ping @tiborvass

**Comment by tiborvass** [comment]
@duglin what would be the ways people could abuse it?

**Comment by duglin** [comment]
@tiborvass can you elaborate? I'm not sure what you're looking for. Not sure how people could "abuse" an environment variable.

**Comment by tiborvass** [comment]
I mean the reason this PR exists is because of PATH=$PATH:/foo/bar would result in :/foo/bar if $PATH was not set initially. I understand the need for ${} in that case, but want to understand what other usecases does ${} affect. Also I wonder whether that usecase alone is worthy of a new feature.

@tianon can you think of anything? Actually, now that I think about it, we should have kept `ENV` only for runtime and never for buildtime. For buildtime environments, just use `RUN export foo=bar` and `ENV foo=$foo` (and make it work).

@duglin to be honest, at this point, you are right, I don't see why we wouldn't support ${}.

**Comment by duglin** [comment]
@tiborvass yes `{$..:-...}` is the primary usecase but I added support for `${...:+...}` too since it was trivial to add - mainly for symmetry with `${...:-...}`. I don't think adding this should be much of a concern since `${}` has been supported for a while and we simply error-out if someone uses a modifier (like - or +) we don't understand. So all we're really doing is changing an error case to a valid case to allow more bash-like semantics. No invention here, just giving the user more stuff that they're used to seeing.

**Comment by tianon** [comment]
More seriously, with what we've already got support for, this does make sense.  It hurts that we're reimplementing so many shell features, but we're already down the hole a ways and can't back out now. :+1:

**Comment by erikh** [comment]
I would much rather see a programmable backend.

Also a gentle reminder I should be pinged on at least builder issues. :)

**Comment by tiborvass** [comment]
@erikh same here, but @duglin's argument to that is that in the mean time we need a fix for these things, plus it's not that much more compared to what we already have.

I'll push this to code review since it seems we agree on this one. Feel free to raise flags if I'm wrong.

**Comment by erikh** [comment]
Nope I've no argument, sorry if I communicated otherwise.

**Comment by duglin** [comment]
@erikh if you get a chance, I wouldn't mind your thoughts on: https://github.com/duglin/docker/tree/Script  cd into 'test' and then build that Dockerfile.

File: builder/shell_parser.go [hunk 1]
```diff
@@ -157,7 +157,38 @@ func (sw *shellWord) processDollar() (string, error) {
 			sw.next()
 			return sw.getEnv(name), nil
 		}
-		return "", fmt.Errorf("Unsupported ${} substitution: %s", sw.word)
+		if ch == ':' {
+			// Special ${xx:...} format processing
+			// Yes it allows for recursive $'s in the ... spot
+			sw.next()
+			modifier := sw.next()
+			if strings.Index("-+", string(modifier)) < 0 {
```
Comment (tiborvass) [inline builder/shell_parser.go]: What about:
```
switch modifier {
case "-":
newValue := sw.getEnv(name)
// ...
case "+":
newValue := sw.getEnv(name)
// ...
default:
// return error
}
```

Comment (duglin) [inline builder/shell_parser.go]: As you wish....

**Comment by tiborvass** [comment]
LGTM

Ping @erikh

**Comment by erikh** [comment]
Just noticing this whole lib now; nice! LGTM.

File: docs/sources/reference/builder.md [hunk 2]
```diff
@@ -113,18 +113,30 @@ images.
...
+used in certain instructions as variables to be interpreted by the
+`Dockerfile`. Escapes are also handled for including variable-like syntax
+into a statement literally.
 
 Environment variables are notated in the `Dockerfile` either with
 `$variable_name` or `${variable_name}`. They are treated equivalently and the
 brace syntax is typically used to address issues with variable names with no
 whitespace, like `${foo}_bar`.
 
+The ${variable_name} syntax also supports a few of the standard `bash`
```
Comment (jamtur01) [inline docs/sources/reference/builder.md]: `\${variable_name}\'

Comment (duglin) [inline docs/sources/reference/builder.md]: @jamtur01 fixed

**Comment by duglin** [comment]
fixed @jamtur01's docs comment

**Comment by jamtur01** [comment]
Docs LGTM

**Comment by duglin** [comment]
is this ready to merge?

**Comment by tiborvass** [comment]
@duglin Yes :)

**Comment by duglin** [comment]
thanks!!

**Comment by thaJeztah** [comment]
🎉😃

**Pull request merged**